File: coding/src/com/algos/li20_dp/_Partitions_L99_UniqueBinarySearchTrees.java
package edu.algos.li20_dp;

public class _Partitions_L99_UniqueBinarySearchTrees {
	public static void main(String[] orgs) {
		_Partitions_L99_UniqueBinarySearchTrees obj = new _Partitions_L99_UniqueBinarySearchTrees();
		
		int n=5;
		System.out.println(obj.numTrees(n));
	}

	public int numTrees(int n) {
		int[] mem = new int[n+1];
		mem[0] = 1;
		mem[1] = 1;
		for(int i=2; i<=n; i++) {
			for(int j=0; j<i; j++) {
				//System.out.println(i+":"+j);
				mem[i] += mem[j]*mem[i-j-1];	// -1 is to exclude the root
			}
		}
		return mem[n];
	}
}

File: dsa/topics/15_dp/python/partitions_l99_unique_binary_search_trees.py
# DP / Catalan Numbers
#
# Problem: Unique Binary Search Trees
# (LeetCode 96)
#
# Idea:
# Number of unique BSTs with n nodes follows Catalan numbers.
# For every node chosen as root, left subtree has j nodes
# and right subtree has (n - j - 1) nodes.
#
# mem[n] = Σ mem[j] * mem[n - j - 1]
#
# Time: O(n^2)
# Space: O(n)

def numTrees(n):
    mem = [0] * (n + 1)
    mem[0] = 1
    mem[1] = 1

    for i in range(2, n + 1):
        for j in range(i):
            # -1 is to exclude the root
            mem[i] += mem[j] * mem[i - j - 1]

    return mem[n]


# -------- Test --------
n = 5
print(numTrees(n))  # Expected: 42
